compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala

```diff
@@ -545,20 +553,16 @@ class ClassfileParser(
       case CLASS_TAG =>
         if (skip) None else Some(lit(Constant(pool.getType(index))))
       case ENUM_TAG =>
-        val enumClassTp = pool.getType(index)
+        val sig = pool.getExternalName(index).value
         val enumCaseName = pool.getName(in.nextChar)
-        if (skip)
-          None
-        else {
-          val enumModuleClass = enumClassTp.classSymbol.companionModule
-          Some(Select(ref(enumModuleClass), enumCaseName.name))
-        }
+        if (skip) None else Some(EnumTag(sig, enumCaseName))
       case ARRAY_TAG =>
-        val arr = new ArrayBuffer[Tree]()
+        val arr = new ArrayBuffer[untpd.Tree]()
         var hasError = false
         for (i <- 0 until index)
           parseAnnotArg(skip) match {
-            case Some(c) => arr += c
+            case Some(c: untpd.Tree) => arr += c
+            case Some(tag: EnumTag) => arr += tag.toTree
```

Comment:
The `toTree` here means we're still forcing things early in some cases

Reply:
I guess if this works OK with JDK 14 we can merge as is but we should revisit it.

Reply:
Yes, this is a recursive call, I see no other ways to deal with it elegantly.

Reply:
We us this call to build up `elems` which is used in `Some(untpd.JavaSeqLiteral(elems, TypeTree()))`, if we want to delay things properly we'd have to propagate the delay to this construction too: replace EnumTag by `Context => untpd.Tree` again, and construct a new `Context => untpd.Tree` when any element in the Array is itself a `Context => untpd.Tree`

Reply:
Yes, if we go back to `Context => untpd.Tree`, things will work, but it will undo the micro-optimizations. I'd say being lazy enough is a good tradeoff. If a problem emerges, we can be lazier (the lazy problem might be in other places as well).